pandas/core/categorical.py

```diff
@@ -316,13 +317,14 @@ def __init__(self, values, categories=None, ordered=None, dtype=None,
             if not isinstance(values, np.ndarray):
                 values = _convert_to_list_like(values)
                 from pandas.core.series import _sanitize_array
-                # On list with NaNs, int values will be converted to float. Use
-                # "object" dtype to prevent this. In the end objects will be
-                # casted to int/... in the category assignment step.
-                if len(values) == 0 or isna(values).any():
+                # By convention, empty lists result in object dtype:
+                if len(values) == 0:
                     sanitize_dtype = 'object'
                 else:
                     sanitize_dtype = None
+                null_mask = isna(values)
+                values = [values[idx] for idx in range(len(values))
```

Comment:
huh? use a vectorized method

Reply:
> huh? use a vectorized method

Notice vectorizing the indexes is the best I could find without transforming the list of values into an array (which must not happen before inferring the dtype).